crdschema/crd.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2022 Upbound Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdschema
+
+import (
+	"os"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/pkg/errors"
+	"github.com/tufin/oasdiff/diff"
+	"github.com/tufin/oasdiff/report"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiyaml "k8s.io/apimachinery/pkg/util/yaml"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+const (
+	errCRDLoad                = "failed to load the CustomResourceDefinition"
+	errBreakingchangesCompute = "failed to compute breaking changes in base and revision CRD schemas"
+)
+
+// Diff can compute schema changes between the base CRD found at `basePath`
+// and the revision CRD found at `revisionPath`.
+type Diff struct {
+	baseCRD     *v1.CustomResourceDefinition
+	revisionCRD *v1.CustomResourceDefinition
+}
+
+// NewDiff returns a new Diff initialized with the base and revision
+// CRDs loaded from the specified base and revision CRD paths.
+func NewDiff(basePath, revisionPath string) (*Diff, error) {
+	d := &Diff{}
+	var err error
+	d.baseCRD, err = getCRD(basePath)
+	if err != nil {
+		return nil, errors.Wrap(err, errCRDLoad)
+	}
+	d.revisionCRD, err = getCRD(revisionPath)
+	if err != nil {
+		return nil, errors.Wrap(err, errCRDLoad)
+	}
+	return d, nil
+}
+
+func getCRD(m string) (*v1.CustomResourceDefinition, error) {
+	crd := &v1.CustomResourceDefinition{}
+	buff, err := os.ReadFile(m)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load the CRD manifest from file: %s", m)
+	}
+	if err := apiyaml.Unmarshal(buff, crd); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal CRD manifest from file: %s", m)
+	}
+	return crd, nil
+}
+
+func getOpenAPIv3Document(crd *v1.CustomResourceDefinition) (*openapi3.T, error) {
+	if len(crd.Spec.Versions) != 1 {
```

Comment:
Thanks @turkenh, I think this is a good point. I believe we can consider extending the functionality if we consider using this tool in other contexts (like moving it to crossplane-tools so that it's more of a generic tool). Currently, we are only planning to use it with Upjet-based providers, which are currently capable of producing a single API version of a CRD. And we would like to use this tool to detect changes in APIs of the same Terraform resource generated for different versions of the underlying Terraform provider.

If, at some point, we would like to generalize the functionality, we can add a new mode for the tool, where we accept a single CRD and detect changes between the different versions declared for that CRD. Do you think we had better introduce a similar functionality now?

Reply:
Added this functionality as a new subcommand. Now, we have:
```shell
crddiff revision <base path> <revision path>
```
to detect changes between two revisions of the same generated CRD (CI usecase), and
```shell
crddiff self <crd path>
```
to detect changes between versions declared in a CRD.

Reply:
Thank you @ulucinar! 
Indeed, I forgot about the primary use case we are focusing on now but great that you've introduced that functionality now.